### PR TITLE
FetchCACert now supports URLs in addition to files

### PR DESCRIPTION
**Pull request by MarcosCela**
**What changed?**

When loading certificates, the function fetchCACert is invoked. Previously, it could only handle
filesystem paths, but now it will also handle external resources if the path is _URL-LIKE_:

If the path starts with _http_, attempt to request and load the response as if it was a certificate.

Certificates for testing have been extracted from:

- https://fm4dd.com/openssl/certexamples.shtm

**Why?**

Sometimes when using the CLI locally, exporting ***$TEMPORAL_CLI_TLS_CA*** to a URL can be useful.
This can solve the problem when certificates are rotated frequently, so you don't have to download
the certificate as a previous step.

**How did you test it?**

- Tested it locally, you can simply host an example server and load the CA by exporting ***$TEMPORAL_CLI_TLS_CA*** .
- Unit tests are also added to identify any possible regression.

**Potential risks**

It's changed behind a conditional:

```go
if strings.HasPrefix(path, "http") {...}
```
Sot it should be pretty safe as far as I am aware.

**Is hotfix candidate?**
No notification required.

Closes: temporalio/temporal#1768

Please let me know what you think!

### CODE REVIEW

File: tools/cli/factory.go [hunk 1]
```diff
@@ -225,11 +237,25 @@ func (b *clientFactory) createTLSConfig(c *cli.Context) (*tls.Config, error) {
...
-func fetchCACert(path string) (*x509.CertPool, error) {
-	caPool := x509.NewCertPool()
-	caBytes, err := ioutil.ReadFile(path)
-	if err != nil {
-		return nil, err
+func fetchCACert(path string) (caPool *x509.CertPool, err error) {
+	caPool = x509.NewCertPool()
+	var caBytes []byte
+
+	if strings.HasPrefix(path, "http") {
```
Comment (lminaudier) [inline tools/cli/factory.go:244]: I think we should rename `path` to `pathOrUrl` and be a little bit more strict with the prefix by checking both `http://` and `https://` so `http.pem` file doesn't get badly interpreted.

Comment (yiminc) [reply tools/cli/factory.go:244]: 👍

**Review by yiminc — APPROVED** [review]

**Review by feedmeapples — APPROVED** [review]

**Comment by sergeybykov** [comment]
Thank you, @MarcosCela!

**Comment by feedmeapples** [comment]
@MarcosCela thx for the contribution! i'm going to be copying this into our updated tctl here https://github.com/temporalio/tctl, if you want to feel free to send a PR there, otherwise i'll just copy with some other updates

File: tools/cli/factory_test.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+	// mock all calls to GET using the testUrl instead
+	return c.client.Get(c.testUrl)
+}
+
+func Test_fetchCACertFromUrl(t *testing.T) {
+	// Ultimately, any URL like:
+	// https://example.com/testdata/4096b-rsa-example-cert.pem
+	// Will attempt to load the local file: testdata/4096b-rsa-example-cert.pem (removing
+	// the host), but "wrapped" in a httptest server
+	tests := []testCase{
```
Comment (mcbryde) [inline tools/cli/factory_test.go:61]: Can we add a test case in which we attempt to pull the cert from a server which itself has an invalid HTTPS certificate?  The current code should error out as written; I want to make sure it continues to.

File: tools/cli/factory.go [hunk 3]
```diff
@@ -225,11 +237,25 @@ func (b *clientFactory) createTLSConfig(c *cli.Context) (*tls.Config, error) {
...
-func fetchCACert(path string) (*x509.CertPool, error) {
-	caPool := x509.NewCertPool()
-	caBytes, err := ioutil.ReadFile(path)
-	if err != nil {
-		return nil, err
+func fetchCACert(pathOrUrl string) (caPool *x509.CertPool, err error) {
+	caPool = x509.NewCertPool()
+	var caBytes []byte
+
+	if strings.HasPrefix(pathOrUrl, "http://") || strings.HasPrefix(pathOrUrl, "https://") {
```
Comment (mcbryde) [inline tools/cli/factory.go:244]: Is there a reason to support HTTP?  If possible we should only support pulling certs over HTTPS.

Comment (feedmeapples) [reply tools/cli/factory.go:244]: https://github.com/temporalio/temporal/pull/1794
сс @MarcosCela

Comment (MarcosCela) [reply tools/cli/factory.go:244]: Thank you @feedmeapples , there was indeed no good reason to allow HTTP!

**Review by mcbryde — COMMENTED** [review]
LGTM with the HTTP change.  Probably should still add that test case for invalid cert on CA server, but not necessary.

**Pull request merged**